### ISSUE TITLE
Why hifigan loss used stft_loss instead of mel_loss?

### ISSUE THREAD

**Issue by hdmjdp** [TensorSpeech/TensorFlowTTS#391]
Hi, Why hifigan loss used stft_loss instead of mel_loss?

**Comment by dathudeptrai** [comment]
@hdmjdp to reuse MelganSTFTTrainer :). I just so lazy, everything is ok :))). STFT loss and Mel-spectrogram loss are almost produced the same performance, it doesn't matter.

**Comment by hdmjdp** [comment]
@dathudeptrai thanks, I will try 48k training using stft_loss. HifiGAN can train g and d together without pretrained generator?

**Comment by dathudeptrai** [comment]
@hdmjdp wow, 48k :)), you should change some param in the training config and preprocessing config. Why don't you try to 22050 Hz first to see the performance :) (same config as ljspeech, you just need to resample ur dataset first.)

**Comment by hdmjdp** [comment]
@dathudeptrai Actually, I have trained the torch version of hifigan on 48k, and it works but train slowly. I think the tf version of yours may train faster than the torch version?

**Comment by dathudeptrai** [comment]
> @dathudeptrai Actually, I have trained the torch version of hifigan on 48k, and it works but train slowly. I think the tf version of yours may train faster than the torch version?

maybe :))). You can try the small config (v2) in this repo first :)).

**Comment by hdmjdp** [comment]
@dathudeptrai I trained G and D together 1k, and I found real_loss and fake_loss is 0. Is that normal? The discriminator_train_start_steps has been seted to 0.

**Comment by dathudeptrai** [comment]
@hdmjdp weird, @ZDisket

**Comment by dathudeptrai** [comment]
@hdmjdp did you check that u use the config that `discriminator_train_start_steps` is 0 ?.

**Comment by hdmjdp** [comment]
@dathudeptrai  step 600 real_loss and fake_loss is not zero,  but later it decrease to 0.
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: gradient_accumulation_steps = 1
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: discriminator_train_start_steps = 0
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: train_max_steps = 2000000
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: save_interval_steps = 20000
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: eval_interval_steps = 5000
2020-11-26 14:00:20,781 (train_hifigan:177) INFO: log_interval_steps = 200
2020-11-26 14:00:20,782 (train_hifigan:177) INFO: num_save_intermediate_results = 1
....


Total params: 58,006,024
Trainable params: 58,006,024
Non-trainable params: 0
_________________________________________________________________
2020-11-26 14:47:52,358 (base_trainer:653) INFO: (Step: 200) train_adversarial_loss = 1.5006.
2020-11-26 14:47:52,359 (base_trainer:653) INFO: (Step: 200) train_fm_loss = 0.0526.
2020-11-26 14:47:52,359 (base_trainer:653) INFO: (Step: 200) train_gen_loss = 11.1889.
2020-11-26 14:47:52,360 (base_trainer:653) INFO: (Step: 200) train_real_loss = 0.0488.
2020-11-26 14:47:52,361 (base_trainer:653) INFO: (Step: 200) train_fake_loss = 0.0087.
2020-11-26 14:47:52,361 (base_trainer:653) INFO: (Step: 200) train_dis_loss = 0.0575.
2020-11-26 14:47:52,362 (base_trainer:653) INFO: (Step: 200) train_spectral_convergence_loss = 5.9100.
2020-11-26 14:47:52,362 (base_trainer:653) INFO: (Step: 200) train_log_magnitude_loss = 4.4626.

**Comment by dathudeptrai** [comment]
> @dathudeptrai step 600 real_loss and fake_loss is not zero, but later it decrease to 0.
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: gradient_accumulation_steps = 1
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: discriminator_train_start_steps = 0
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: train_max_steps = 2000000
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: save_interval_steps = 20000
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: eval_interval_steps = 5000
> 2020-11-26 14:00:20,781 (train_hifigan:177) INFO: log_interval_steps = 200
> 2020-11-26 14:00:20,782 (train_hifigan:177) INFO: num_save_intermediate_results = 1
> ....
> 
> Total params: 58,006,024
> Trainable params: 58,006,024
> Non-trainable params: 0
> 
> 2020-11-26 14:47:52,358 (base_trainer:653) INFO: (Step: 200) train_adversarial_loss = 1.5006.
> 2020-11-26 14:47:52,359 (base_trainer:653) INFO: (Step: 200) train_fm_loss = 0.0526.
> 2020-11-26 14:47:52,359 (base_trainer:653) INFO: (Step: 200) train_gen_loss = 11.1889.
> 2020-11-26 14:47:52,360 (base_trainer:653) INFO: (Step: 200) train_real_loss = 0.0488.
> 2020-11-26 14:47:52,361 (base_trainer:653) INFO: (Step: 200) train_fake_loss = 0.0087.
> 2020-11-26 14:47:52,361 (base_trainer:653) INFO: (Step: 200) train_dis_loss = 0.0575.
> 2020-11-26 14:47:52,362 (base_trainer:653) INFO: (Step: 200) train_spectral_convergence_loss = 5.9100.
> 2020-11-26 14:47:52,362 (base_trainer:653) INFO: (Step: 200) train_log_magnitude_loss = 4.4626.

i think you should follow my instruction. Training 100k steps then training Discriminator + generator.

**Comment by hdmjdp** [comment]
@dathudeptrai That's wired, the torch version no need pretrained 100k.

**Comment by dathudeptrai** [comment]
> @dathudeptrai That's wired, the torch version no need pretrained 100k.

the training process is different :D. But i do not know, ur spectral + log loss should be < 1.0 after around 3k-5k steps.

**Comment by hdmjdp** [comment]
@dathudeptrai thanks，i will try to refine code to align tf version with torch version.

**Comment by dathudeptrai** [comment]
> @dathudeptrai thanks，i will try to refine code to align tf version with torch version.

no need to align :))) i finish my hifigan in my private library and that ok.

**Comment by hdmjdp** [comment]
@dathudeptrai Does your private library version train 100k steps for generator only then training Discriminator + generator?
Have you train g and d from scratch?

**Comment by dathudeptrai** [comment]
> @dathudeptrai Does your private library version train 100k steps for generator only then training Discriminator + generator?
> Have you trained g and d from scratch?

pretrained G before training both always better.

**Comment by hdmjdp** [comment]
@dathudeptrai Ok, got it. thanks.
